### PR TITLE
Support SinceTs in iterators

### PR DESCRIPTION
**Pull request by jarifibrahim**
SinceTs can be used to read data above a particular timestamp. All data
less with version less than the sinceTs will be ignored.

SinceTs is always less than the readTs and when sinceTs is set, the
data between sinceTs and readTs will be read.

sinceTs < data to be read <= readTs

Fixes DGRAPH-2958

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1653)

### CODE REVIEW

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1653)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @jarifibrahim and @manishrjain)

___
*[iterator.go, line 399 at r1](https://reviewable.io/reviews/dgraph-io/badger/1653#-MRtlDOR3kToyuQWb2R3:-MRtlDOR3kToyuQWb2R4:b9ubcig) ([raw file](https://github.com/dgraph-io/badger/blob/93260b6fd605a2ed7a83063e44cd1f014b3b537d/iterator.go#L399)):*
> ```Go
> 	hash := y.Hash(opt.Prefix)
> 	for _, t := range filtered {
> 		if t.MaxVersion() < opt.SinceTs {
> ```

If the prefix is set to `nil` then we will end up picking all the tables.
___
*[iterator_test.go, line 149 at r1](https://reviewable.io/reviews/dgraph-io/badger/1653#-MRtnEl839QGzbQw8fjx:-MRtnEl839QGzbQw8fjy:b-gqztty) ([raw file](https://github.com/dgraph-io/badger/blob/93260b6fd605a2ed7a83063e44cd1f014b3b537d/iterator_test.go#L149)):*
> ```Go
> 		iopt.SinceTs = sinceTs
> 
> 		db.View(func(txn *Txn) error {
> ```

require.NoError()

**Review by NamanJain8 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1653)* status: 0 of 5 files reviewed, 1 unresolved discussion (waiting on @jarifibrahim and @manishrjain)

**Review by ahsanbarkati — APPROVED** [review]

File: backup.go [hunk 1]
```diff
@@ -68,11 +69,6 @@ func (stream *Stream) Backup(w io.Writer, since uint64) (uint64, error) {
 			if !bytes.Equal(item.Key(), key) {
 				return list, nil
 			}
-			if item.Version() < since {
```
Comment (jarifibrahim) [inline backup.go:71]: log an error if this happens.

File: iterator.go [hunk 2]
```diff
@@ -362,34 +367,42 @@ func (opt *IteratorOptions) pickTable(t table.TableInterface) bool {
 // pickTables picks the necessary table for the iterator. This function also assumes
 // that the tables are sorted in the right order.
 func (opt *IteratorOptions) pickTables(all []*table.Table) []*table.Table {
+	allCopy := make([]*table.Table, len(all))
+	copy(allCopy, all)
+	if opt.SinceTs > 0 {
+		tmp := allCopy[:0]
```
Comment (jarifibrahim) [inline iterator.go:373]: This filtering should be the last step in pickTables.

File: iterator.go [hunk 3]
```diff
@@ -610,7 +623,8 @@ func (it *Iterator) parseItem() bool {
 
 	// Skip any versions which are beyond the readTs.
 	version := y.ParseTs(key)
-	if version > it.readTs {
+	// Ignore everything that is above the readTs and below the sinceTs.
+	if version > it.readTs || version < it.opt.SinceTs {
```
Comment (jarifibrahim) [inline iterator.go:627]: Don't include the data at the sinceTs.

**Pull request merged**